### PR TITLE
fixed memory leak in cSatipSectionFilter

### PR DESCRIPTION
**Pull request by pipelka**
This fixes a rather nasty one.
valgrind is your friend ;-)

### CODE REVIEW

**Comment by rofafor** [comment]
The indentation of modified code is off.

Can you explain this change a bit more as I don't fully understand why we should always drop sections as this leads to data loss when VDR's section handler is temporarily overloaded. At a quick glance it seems to me that your VDR cannot do section processing in time and therefore the plugin starts queuing sections that might look like a memory leak.

**Comment by pipelka** [comment]
The question here is: What happens to frames (sections) when send() fails ?
valgrind definitely claims that these frames are indirectly lost (without the patch).

**Comment by rofafor** [comment]
The failed section frame should stay in the ring buffer and the section filter handler thread should retry them. If the ring buffer gets full, any new sections should be discarded - unfortunately silently as there are currently no overflow detection:
```
int cSatipSectionFilter::Filter(void):

     if (ringBufferM && (secLenM > 0))
-        ringBufferM->Put(new cFrame(secBufM, secLenM));
+        if (!ringBufferM->Put(new cFrame(secBufM, secLenM))) error("Filter overrun!");
```
Now, that Valgrind states that there are blocks indirectly lost, there's a bug somewhere (maybe even in the ringbuffer class itself. Have you tried a bit modern approach of using clang's LeakSanitizer?

**Comment by rofafor** [comment]
From VDR documentation:
```
 cFrame *Get(void);
    // Gets the next frame from the ring buffer.
    // The actual data still remains in the buffer until Drop() is called.
  void Drop(cFrame *Frame);
    // Drops the Frame that has just been fetched with Get().
```
The code ```cFrame *section = ringBufferM->Get();``` should be quite safe to execute as many time as possible, but I guess Valgrind detects that as a memory leak via a lost pointer. So I still think it's just a false positive. Have you verified the leak from process statistics?

**Comment by pipelka** [comment]
I started searching for leaks as memory statistics showed increasing values. But it's quite hard to argue from process statistics as vdr eats up a lot of memory.
I'll do more debugging later in the evening, ...

**Comment by pipelka** [comment]
The funny thing is that remaining cFrame objects should be cleared by the ringbuffer destructor.
Obviously this doesn't happen. I'm asking myself: Why ?
From my experience valgrind is always right. I thought countless times about false positive leaks, until I found the reason, ...

**Comment by rofafor** [comment]
Now that the sections aren't crucial ones, one might just drop them from time to time, but the ringbuffer available check should be done after the drop. Now, if we are dropping every section frame, there's no need for retry logic either and that that should be ripped off as well.

**Comment by rofafor** [comment]
What are your compiling options? Does your compiler optimize the code too aggressively?

**Comment by pipelka** [comment]
It's "-O3 -g" with gcc.

**Comment by pipelka** [comment]
Ok. Curing symptoms till now ;-)
I found the source of the evil.
valgrind is clean with the modified patch.

**Comment by pipelka** [comment]
I would also suggest to replace the send & retry logic with using poll() and send(). Additionally there sould by a timeout sending data. Currently the Send() method assumes the all data is transferred in one swoop. In reality partial transfers happen regurarely so this should also be taken into account.

File: sectionfilter.c [hunk 1]
```diff
@@ -102,8 +102,11 @@ int cSatipSectionFilter::Filter(void)
      if (doneqM && !neq)
         return 0;
 
-     if (ringBufferM && (secLenM > 0))
-        ringBufferM->Put(new cFrame(secBufM, secLenM));
+     if (ringBufferM && (secLenM > 0)) {
+        cFrame* section = new cFrame(secBufM, secLenM);
+        if (!ringBufferM->Put(section))
+           delete section;
```
Comment (rofafor) [inline sectionfilter.c]: Please, change this to ```DELETE_POINTER(section);```

**Comment by pipelka** [comment]
Ok. DELETE_POINTER is in.

**Comment by rofafor** [comment]
Great. Could you reword your commit message to start with a capital letter and end with a period.

IIRC, I couldn't get any more elegant Unix socket implementation working correctly with the core VDR, but I might have made mistakes back then...

**Comment by pipelka** [comment]
Alright. Commit message changed.
Anything else, jump throug a loop or something, ... Just kidding ;-)

**Comment by rofafor** [comment]
Heh. Sorry for nitpicking, but I just tried to maintain the existing coding style. Bit thanks for your great contribution. Thanks!

**Comment by pipelka** [comment]
No problem at all. ;-)
Next PR will arrive soon, ...

**Pull request merged**